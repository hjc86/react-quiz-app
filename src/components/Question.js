import React, { Component } from 'react'
import ReactDOM from "react-dom"
import Answer from './Answer'
import '../App.css';
import Score from './Score';
import styles from './Question.module.css';

class Question extends Component {
    constructor(props){
        super(props)
        this.state={
            randomisedAnswers:null,
            correctAnswer: null,
            wrongAnswers: null,
            backgroundStyle: "none",
            clicked: false,
            userAnswers : [],
            questionDecoded: ''
        }
    }

    createAnswerArray= ()=> {  
        let questionData= this.props.questionData 
        let answerArray= questionData.incorrect_answers.concat([questionData.correct_answer])
        let arr = this.getRandomArray(answerArray)
        this.setState(
            {
                randomisedAnswers: arr,
                correctAnswer: questionData.correct_answer,
                wrongAnswers: questionData.incorrect_answers,
            }
        )
    }

    getRandomArray= (array) => {
        const questionAnswerIndicies = []
        const randomisedAnswers = []
        while(questionAnswerIndicies.length<array.length){
           let randomIndex = Math.floor(Math.random() * Math.floor(array.length));
            if(!questionAnswerIndicies.includes(randomIndex)){
                questionAnswerIndicies.push(randomIndex)
                randomisedAnswers.push(array[randomIndex])
            };
        }
         
        let randomisedAnswersDecoded= randomisedAnswers.map((answer)=> {
            let txt = document.createElement("textarea")
            txt.innerHTML = answer
            return txt.value
        })
            
        return randomisedAnswersDecoded
    }

    hasBeenClicked = (answerData)=>{
        this.setState({
            clicked : true,
            selectedAnswer: answerData.selectedAnswer,
            backgroundStyle: answerData.backgroundStyle
        })

        if(answerData.addPoint){
            this.props.latestPoint(1)
        }
        else{
            this.props.latestPoint(0)
        }

        this.props.answeredQuestions(1)
    }

    decodeHtml(html) {
        var txt = document.createElement("textarea");
        txt.innerHTML = html;
        this.setState({
            questionDecoded: txt.value
        })
    }

    componentDidMount(){
        this.createAnswerArray()
        this.decodeHtml(this.props.questionData.question)
    }


    render() {
                return (
                    
                    <div className={styles.questionGroup}>

                    {this.state.randomisedAnswers === null ? "waiting for quiz to be generated" :
                        <React.Fragment>
        
                            <div className={styles.questionTitle}>
        

                                <div className={styles.questionBox}>
                                            
                                    {this.state.questionDecoded}

                                </div>
                            </div>
                            {this.state.randomisedAnswers.map((answer)=> (
                            <Answer {...this.state} answer={answer} clicked={this.hasBeenClicked} className={this.state.clicked ? "inactive": ""} />
                            ))}
                        </React.Fragment>
                    }
                       </div>
                )
            }

}
export default Question