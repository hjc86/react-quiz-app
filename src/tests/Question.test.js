import React, { Component } from 'react';
import ReactDOM from 'react-dom'
import { shallow, mount } from 'enzyme';
import Question from '../components/Question';
import renderer from 'react-test-renderer';

let testData ={ 
    
        "category": "Animals",
        "type": "multiple",
        "difficulty": "easy",
        "question": "Hippocampus is the Latin name for which marine creature?",
        "correct_answer": "Seahorse",
        "incorrect_answers": [
          "Dolphin",
          "Whale",
          "Octopus"
        ]
    }


describe("<Question />", () => {
    
    /// setup for test ///

    const essentialProps ={
        questionData: testData,
        latestPoint: 0,
        answeredQuestions: 0
    }

    let componentWithProps = <Question {...essentialProps}/>
    let wrapper;

    /// end of setup //

    beforeEach(() => wrapper = shallow(componentWithProps));

    it('should render  three <div />',()=>{
        expect(wrapper.find('div').length).toEqual(3);
    })

    it('mounts to DOM', () => {
        const div = document.createElement('div');
        ReactDOM.render(componentWithProps, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(componentWithProps)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    it('Renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })


})
