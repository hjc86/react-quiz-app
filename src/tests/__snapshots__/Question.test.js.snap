// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Question /> Renders correctly 1`] = `ShallowWrapper {}`;

exports[`<Question /> renders correctly 1`] = `
<div>
  <div
    style={
      Object {
        "border": "5px solid black",
      }
    }
  >
    <div
      className="questionAnswers"
    >
      Hippocampus is the Latin name for which marine creature?
    </div>
  </div>
  <div
    className="buttonDiv"
  >
    <button
      className="answers "
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "none",
          "display": undefined,
        }
      }
    >
      Seahorse
    </button>
  </div>
  <div
    className="buttonDiv"
  >
    <button
      className="answers "
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "none",
          "display": undefined,
        }
      }
    >
      Dolphin
    </button>
  </div>
  <div
    className="buttonDiv"
  >
    <button
      className="answers "
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "none",
          "display": undefined,
        }
      }
    >
      Whale
    </button>
  </div>
  <div
    className="buttonDiv"
  >
    <button
      className="answers "
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "none",
          "display": undefined,
        }
      }
    >
      Octopus
    </button>
  </div>
</div>
`;
